wifi-bluetooth-support.robot: nmcli package check & installation

matmacieje opened the pull request:
network-manager package check added in WLE002.001 test case

DonVasyl committed:
WLE002.001: network-manager & pciutil check & installation for Ubuntu

wifi-bluetooth-support.robot, WLE002.001:
 - dhclient call added before Detect Or Install Package
 - Ubuntu-specific package installation moved from test case
 to WiFi Scanning kw

keywords.robot:
 - added apt list & regex-based package detection in
 Check If Package Is Installed
 - Install Package redone using Execute Command In Terminal

lib/network.robot:
 - Execute Command In Terminal used instead of Execute Linux Command
in Check Internet Connection On Linux keyword

lib/telnet.robot:
 - Telnet.Read added in Execute Command In Terminal

Signed-off-by: Mateusz Maciejewski <mateusz.maciejewski@3mdeb.com>

macpijan commented:
Has it been tested? Can you show logs? It shows that `network-manager` is installed, but `nmcli` is still unavailable?


```
WLE002.001 Wi-Fi scanning (Ubuntu) :: Check whether the Wi-Fi func... .....
Checking if network-manager is installed...

Package network-manager is installed
WLE002.001 Wi-Fi scanning (Ubuntu) :: Check whether the Wi-Fi func... | FAIL |
No match found for '3mdeb_Laboratorium' in 1 minute. Output:
snap install network-manager  # version 1.2.2-31, or
apt  install network-manager  # version 1.46.0-1ubuntu2.2
See 'snap info network-manager' for additional versions.
root@3mdeb:/home/ubuntu# nmcli device wifi rescan
nmcli device wifi list
Command 'nmcli' not found, but can be installed with:
snap install network-manager  # version 1.2.2-31, or
apt  install network-manager  # version 1.46.0-1ubuntu2.2
See 'snap info network-manager' for additional versions.
root@3mdeb:/home/ubuntu# nmcli device wifi list
Command 'nmcli' not found, but can be installed with:
snap install network-manager  # version 1.2.2-31, or
apt  install network-manager  # version 1.46.0-1ubuntu2.2
See 'snap info network-manager' for additional versions.
root@3mdeb:/home/ubuntu#
```

matmacieje commented:
@macpijan yes, it is tested, here's the log.

[WLE_nmcli.zip](https://github.com/user-attachments/files/19531377/WLE_nmcli.zip)

I've also checked for differences between Ubuntu versions, if nmcli binary is being provided by another package in the past.

miczyg1 commented:
> I've also checked for differences between Ubuntu versions, if nmcli binary is being provided by another package in the past.

https://manpages.ubuntu.com/manpages/focal/man1/nmcli.1.html

It says that the nmcli is provided by network-manager package since Ubuntu 20.04LTS and it doesn' change for any newer version.

> Has it been tested? Can you show logs? It shows that network-manager is installed, but nmcli is still unavailable?

This actually shows that the package is not installed:

```
root@3mdeb:/home/ubuntu# nmcli device wifi rescan
nmcli device wifi list
Command 'nmcli' not found, but can be installed with:
snap install network-manager  # version 1.2.2-31, or
apt  install network-manager  # version 1.46.0-1ubuntu2.2
See 'snap info network-manager' for additional versions.
```

Or that `nmcli` was somehow erased without removing the `network-manager` package.

miczyg1 commented:
More context:

`dpkg --list network-manager` will give an output:

```
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name            Version           Architecture Description
+++-===============-=================-============-============================>
rc  network-manager 1.46.0-1ubuntu2.2 amd64        network management framework>
```

Even if the package is not installed...
Confirming with apt is the way:

```
apt list --installed 2> /dev/null |grep network-manager
```

Gives empty output. So the keyword `Check If Package Is Installed` is broken. If I install network manager and repeat, I get:

```
apt list --installed 2> /dev/null |grep network-manager
network-manager-pptp/noble,now 1.2.12-3build2 amd64 [installed,automatic]
network-manager/noble-updates,now 1.46.0-1ubuntu2.2 amd64 [installed]
```

miczyg1 reviewed (changes requested):
Still does not work reliably. See my comment: https://github.com/Dasharo/open-source-firmware-validation/pull/746#issuecomment-2765942352

matmacieje commented:
Please review, log in attachment.

[WLE002.001.zip](https://github.com/user-attachments/files/19537517/WLE002.001.zip)

miczyg1 commented on keywords.robot:
```diff
@@ -1173,6 +1173,34 @@ Install Package
     Read From Terminal Until Prompt
     Set DUT Response Timeout    180s
 
+Reinstall Package
+    [Documentation]    Keyword allows to reinstall the package, that is
+    ...    necessary to run the test case. Use when package contents are
+    ...    missing.
+    [Arguments]    ${package}
+    Wait Until Keyword Succeeds    10x    5s
+    ...    Check Internet Connection On Linux
+    Set DUT Response Timeout    600s
+    Write Into Terminal    apt-get update
+    Read From Terminal Until Prompt
+    Write Into Terminal    apt-get install --reinstall --assume-yes ${package}
+    Read From Terminal Until Prompt
+    Set DUT Response Timeout    180s
+
+Check Command Binary
+    [Documentation]    The keyword verifies, if command binary
+    ...    exists on chosen path using whereis -b command.
+    [Arguments]    ${bin_name}    ${bin_name_with_path}
+    Write Into Terminal    whereis -b ${bin_name}
+    ${output}=    Read From Terminal Until Prompt
+    ${regex_result}=    Get Lines Matching Pattern    ${output}    ${bin_name_with_path}
+    IF    "${regex_result}"!="${EMPTY}"
+        Set Test Variable    ${RESULT}    True
+    ELSE
+        Set Test Variable    ${RESULT}    False
+    END
+    RETURN    ${RESULT}
```
Why reinstall and check command presence? Simply fix `Check If Package Is Installed` keyword like I explained https://github.com/Dasharo/open-source-firmware-validation/pull/746#issuecomment-2765942352

matmacieje replied:
Removed in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/c30ea0c62099738ea8256936a74b9db1f72b93da

matmacieje commented:
Please review, logs in attachment. Both required packages (network-manager & pciutils) were 'apt remove'd for this execution.

[WLAN_NM_regex_vp6650.zip](https://github.com/user-attachments/files/19565390/WLAN_NM_regex_vp6650.zip)

miczyg1 commented on dasharo-compatibility/wifi-bluetooth-support.robot:
```diff
@@ -73,6 +73,11 @@ WLE002.001 Wi-Fi scanning (Ubuntu)
+    # check_internet_connection set to False
+    # we either have cached package or working internet connection
+    # that is not NM-controlled
+    Detect Or Install Package    network-manager    False
```
Why we cannot check the connection here? How do we expect to install the package if it is missing and the network connection is not ready?

matmacieje replied:
After "apt remove network-manager" and DUT Power On or reboot there is no internet access on default Ubuntu installation. If package network-manager was removed and not purged, this works offline. There may be also Ubuntu systems with internet connection without NM or with connection not controlled by NM.

miczyg1 replied:
Simply invoking `sudo dhclient` helped me in this case, because the interfaces where not up automatically and did not have an IP address, that's all. Why not execute this command before `Detect Or Install Package` ?

Modifying a generic keyword is not usually the way to go.

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/639b66dedcdbcba24edfc1627e18fa892690e29d

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/b547f37e05ed5353fdfbda27f3fc75b58f76b0eb

miczyg1 commented on keywords.robot:
```diff
@@ -1156,8 +1158,16 @@ Check If Package Is Installed
-    IF    "no packages found matching" in """${output}""" or "<none>" in """${output}""" or "dpkg was interrupted" in """${output}"""
+
+    # read from terminal, to be sure that there will be no previous command output captured
+    Read From Terminal
```
THere shouldn't be any output from previous commands. The keywords are designed the consume the output. What exactly was the leftover from previous command here?

matmacieje replied:
It was either apt output or nmcli output.

[wifi-bluetooth-support.robot_log.zip](https://github.com/user-attachments/files/19566761/wifi-bluetooth-support.robot_log.zip)

miczyg1 replied:
Thank you for the log. Now I can see why. So:

1. `Check Internet Connection On Linux` uses the wrong keyword `Execute Linux Command` instead of `Execute Command In Terminal`. 
2. `Install Package` uses the wrong keyword `Write Into Terminal` instead of `Execute Command In Terminal`. 

Generally it looks like a chaos. `Execute Linux Command` is still used in many places, where it shouldn't. AFAIK `Execute Command In Terminal` is the keyword to go when executing commands in an OS with a prompt. Was there any cleanup made to use `Execute Command In Terminal` everywhere (where applicable) @macpijan ?

We won't fix everything here, so at least if we fix those 2 points I mentioned above @matmacieje , we should be able to remove `Read From Terminal` and close this thread.

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/b547f37e05ed5353fdfbda27f3fc75b58f76b0eb

miczyg1 commented on keywords.robot:
```diff
@@ -1156,8 +1158,16 @@ Check If Package Is Installed
+
+    # there's limitation (or possibly a bug) in Get Lines Matching Regexp kw:
+    # can't capture opening square bracket correctly, which would be more appropriate
+    ${package_regex}=    Catenate    SEPARATOR=    ${package}    \/.*installed.*\]
```
Brackets are special signs in regexp, no wonder it causes problems. `\/.*installed.*` should be enough here.

matmacieje replied:
Should I remove` \]  `?

miczyg1 replied:
Yes and recheck

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/639b66dedcdbcba24edfc1627e18fa892690e29d

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/b547f37e05ed5353fdfbda27f3fc75b58f76b0eb

miczyg1 commented on keywords.robot:
```diff
@@ -1169,6 +1179,9 @@ Install Package
     [Arguments]    ${package}
     Set DUT Response Timeout    600s
+
+    Write Into Terminal    apt-get update
+    Read From Terminal Until Prompt
     Write Into Terminal    apt-get install --assume-yes ${package}
     Read From Terminal Until Prompt
```
Why this keyword doesn't use `Execute Command In Terminal`?

matmacieje replied:
I tested both variants, but decided to keep it in original form, to reduce number of changes. Do you want this to be changed to Execute Command In Terminal?

miczyg1 replied:
Yes, see my comment https://github.com/Dasharo/open-source-firmware-validation/pull/746#discussion_r2024880443

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/b547f37e05ed5353fdfbda27f3fc75b58f76b0eb

matmacieje commented:
Please review, rebase to develop done, here's log: 
[wifi_nm_pciutils.zip](https://github.com/user-attachments/files/19590254/wifi_nm_pciutils.zip)

matmacieje commented:
Rebase. 
[wifi-bluetooth-support_2025_04_14_12_05_57.zip](https://github.com/user-attachments/files/19733767/wifi-bluetooth-support_2025_04_14_12_05_57.zip)

miczyg1 commented on dasharo-compatibility/wifi-bluetooth-support.robot:
```diff
@@ -163,8 +163,16 @@ Wi-Fi Scanning
     Switch To Root User
-    Scan For Wi-Fi In Linux
 
+    # with interfaces DOWN, dhclient takes around 37s
+    Execute Command In Terminal    dhclient    60s
+    IF    "${BOOTED_OS_ID}"=="${ENV_ID_UBUNTU}"
+        Detect Or Install Package    network-manager
+    END
+    Scan For Wi-Fi In Linux
+    IF    "${BOOTED_OS_ID}"=="${ENV_ID_UBUNTU}"
+        Detect Or Install Package    pciutils
+    END
```
It doesn't hurt to check for packages here, but I heard that we extend the ansible now.

See an example here:

https://github.com/Dasharo/open-source-firmware-validation/blob/develop/os-config/ansible/linux-packages-playbook.yaml

Could you please extend it with required packages?

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/0d35a2e2b1c09e44cf3d5d514a1368a8005dcc6c

miczyg1 commented on keywords.robot:
```diff
@@ -1155,8 +1155,13 @@ Check If Package Is Installed
-    IF    "no packages found matching" in """${output}""" or "<none>" in """${output}""" or "dpkg was interrupted" in """${output}"""
+    ${apt_list_output}=    Execute Command In Terminal    apt list --installed 2> /dev/null | grep ${package}    60s
+
+    # there's limitation (or possibly a bug) in Get Lines Matching Regexp kw:
+    # can't capture opening square bracket correctly, which would be more appropriate
```
It is not a bug. Brackets are special characters in regexp. We can remove this comment, it does not bring any valuable info.

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/746/commits/0d35a2e2b1c09e44cf3d5d514a1368a8005dcc6c

DonVasyl committed:
WLE002.201 & .202: required packages are now also in ansible playbook

- network-manager & pciutils are now installed by ansible playbook
- regexp comment removed

Signed-off-by: Mateusz Maciejewski <mateusz.maciejewski@3mdeb.com>

matmacieje commented:
Please review, logs in attachment:
[basic-platform-setup_2025_04_14_15_55_28.zip](https://github.com/user-attachments/files/19736912/basic-platform-setup_2025_04_14_15_55_28.zip)
[wifi-bluetooth-support_2025_04_14_16_03_17.zip](https://github.com/user-attachments/files/19736913/wifi-bluetooth-support_2025_04_14_16_03_17.zip)
